internal/concurrent/goroutine_limiter.go

```diff
@@ -0,0 +1,32 @@
+package concurrent
+
+import (
+	"context"
+
+	"golang.org/x/sync/semaphore"
+)
+
+type LimitedGoRoutineRunner struct {
+	sem *semaphore.Weighted
+}
+
+// NewGoroutineLimiter creates a new GoroutineRunner that limits the number of goroutines. No more than the limit number
+// of goroutines can be running at the same time.
+func NewGoroutineLimiter(limit int64) *LimitedGoRoutineRunner {
+	return &LimitedGoRoutineRunner{
+		sem: semaphore.NewWeighted(limit),
+	}
+}
+
+func (l *LimitedGoRoutineRunner) Go(ctx context.Context, fn func()) error {
```

Comment:
Should document out that this will block rather surprisingly in the middle of NewFuture rather than in Future.Get. Is that desirable -- I guess it avoids a proliferation of errors but is a pretty big break to the contract of NewFuture

Reply:
Good point about documentation! 

This behavior is desired albeit unclear)! The point of the GoRoutineLimiter is to limit the number of go routines we spin up. It also lets us force concurrent code to be serializable to make our schema fetching compatible with connections.

Reply:
> The point of the GoRoutineLimiter is to limit the number of go routines we spin up
Is our goal to limit the number of goroutines or to limit the number of goroutines doing the expensive work? 

> It also lets us force concurrent code to be serializable to make our schema fetching compatible with connections.
Wouldn't this happen anyways if you moved the semaphore check inside the goroutine execution? It would occur during Future.Get, which I figure is the expected behavior.

Reply:
It's too limit the number of go routines. Go supports hundreds of thousands of go routines, but it's a bad practice to leave it entirely unbounded.

> Wouldn't this happen anyways if you moved the semaphore check inside the goroutine execution? It would occur during Future.Get, which I figure is the expected behavior.
No, it wouldn't without a specialized future implementation, since futures eagerly compute. 

We would inherently be spinning up  goroutines if we didn't block on this aspect, which makes the code not thread-safe. For instance, I tried making the code serializable initially by setting go routine limit == 1, but that doesn't work because if a future calls into another a future, the shared "thread pool" will have deadlock, since 1 go routine can't complete with the other.